server/etcdserver/apply/apply_auth_test.go

```diff
@@ -15,47 +15,844 @@
+				request:               &pb.InternalRaftRequest{AuthRoleDelete: &pb.AuthRoleDeleteRequest{}},
+				adminPermissionNeeded: true,
+			},
+			{
+				name:                  "AuthRoleGrantPermission needs admin permission",
+				request:               &pb.InternalRaftRequest{AuthRoleGrantPermission: &pb.AuthRoleGrantPermissionRequest{}},
+				adminPermissionNeeded: true,
+			},
+			{
+				name:                  "AuthRoleRevokePermission needs admin permission",
+				request:               &pb.InternalRaftRequest{AuthRoleRevokePermission: &pb.AuthRoleRevokePermissionRequest{}},
+				adminPermissionNeeded: true,
+			},
+		}
+	authApplier := defaultAuthApplierV3(t)
+	MustCreateRolesAndEnableAuth(t, authApplier)
+	dummyApplyFunc := func(ctx context.Context, r *pb.InternalRaftRequest, shouldApplyV3 membership.ShouldApplyV3) *Result {
+		return &Result{}
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			if tc.adminPermissionNeeded {
+				tc.request.Header = &pb.RequestHeader{Username: userReadOnly}
+			}
+			result := authApplier.Apply(ctx, tc.request, false, dummyApplyFunc)
+			require.Equal(t, result.Err == auth.ErrPermissionDenied, tc.adminPermissionNeeded,
+				"Admin permission needed: got %v, expect: %v", result.Err == auth.ErrPermissionDenied, tc.adminPermissionNeeded)
+		})
+	}
+}
+
+// TestAuthApplierV3_Put verifies only users with write permissions in the key range can put
+func TestAuthApplierV3_Put(t *testing.T) {
+	tcs := []struct {
+		name           string
+		userName       string
+		request        *pb.PutRequest
+		expectError    bool
+		expectErrorMsg string
+	}{
+		{
+			name:           "put permission denied",
+			userName:       userReadOnly,
+			request:        &pb.PutRequest{},
+			expectError:    true,
+			expectErrorMsg: auth.ErrPermissionDenied.Error(),
+		},
+		{
+			name:     "prevKv is set, but user does not have read permission",
+			userName: userWriteOnly,
+			request: &pb.PutRequest{
+				Key:    []byte(key),
+				Value:  []byte("1"),
+				PrevKv: true,
+			},
+			expectError:    true,
+			expectErrorMsg: auth.ErrPermissionDenied.Error(),
+		},
+		{
+			name:     "put success",
+			userName: userWriteOnly,
+			request: &pb.PutRequest{
+				Key:   []byte(key),
+				Value: []byte("1"),
+			},
+			expectError: false,
+		},
+		{
+			name:     "put success with PrevKv set",
+			userName: userRoot,
+			request: &pb.PutRequest{
+				Key:    []byte(key),
+				Value:  []byte("1"),
+				PrevKv: true,
+			},
+			expectError: false,
+		},
+	}
+
+	authApplier := defaultAuthApplierV3(t)
+	MustCreateRolesAndEnableAuth(t, authApplier)
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			SetAuthInfo(authApplier, tc.userName)
+			_, _, err := authApplier.Put(ctx, nil, tc.request)
+			if (err != nil) != tc.expectError {
```

Comment:
you can simplify the two if's by just checking whether there was an error:

> require.Equal(t, tc.expectError, err)

it will handle nil's just fine, no need to dance around with the booleans. You can also equality check the error itself or leave it as nil when there's no error.